README.md
# emailee

[![PyPI Version][pypi-image]][pypi-url]
[![Build Status][build-image]][build-url]
[![Code Coverage][coverage-image]][coverage-url]
[![Code Quality][quality-image]][quality-url]
[![Code Factor][codefactor-image]][codefactor-url]
[![versions][versions-image]][versions-url]
[![license][license-image]][license-url]

<!-- Badges -->

[pypi-image]: https://img.shields.io/pypi/v/emailee
[pypi-url]: https://pypi.org/project/emailee/

[build-image]: https://github.com/mattwalshdev/emailee/actions/workflows/build.yml/badge.svg
[build-url]: https://github.com/mattwalshdev/emailee/actions/workflows/build.yml

[coverage-image]: https://codecov.io/gh/mattwalshdev/emailee/branch/main/graph/badge.svg
[coverage-url]: https://codecov.io/gh/mattwalshdev/emailee

[quality-image]: https://api.codeclimate.com/v1/badges/0360d77f65da3d985d03/maintainability
[quality-url]: https://codeclimate.com/github/mattwalshdev/emailee

[codefactor-image]: https://www.codefactor.io/repository/github/mattwalshdev/emailee/badge
[codefactor-url]: https://www.codefactor.io/repository/github/mattwalshdev/emailee

[versions-image]: https://img.shields.io/pypi/pyversions/emailee.svg
[versions-url]: https://github.com/mattwalshdev/emailee

[license-image]: https://img.shields.io/github/license/mattwalshdev/emailee.svg
[license-url]: https://github.com/mattwalshdev/emailee/blob/main/LICENSE

A robust SMTP mailer library, with async mail send features built-in using Threads and Multiprocessing. Zero PyPI (production) dependencies.

## Installation

Install with `pip install emailee`.

## Some quick examples

Send a simple text email

```Python
import emailee

my_email = emailee.Emailee()
my_email.sender("john.smith@fakeemail.com")
my_email.subject("Test email")
my_email.msgContent("some poorly formatted raw text")
my_email.sendTo(["jill.smith@fakeemail.com"])
my_email.server("smtp.fakeemail.com")

if my_email.ready():
    my_email.send()
```

Send multiple asyncronous emails

```Python
import emailee

emails_list = [{...}]
server_dict = {...}

emails = emailee.AsyncThreads(emails_list, server_dict, outputFile='output.txt')
```

## Documentation

There are only three classes exposed to the user of this library, `Emailee`, `AsyncThreads` and `AsyncMP`. The latter two utilise the Emailee class to send emails via the Threading or Multiprocessing async APIs respectively.

### Emailee class

If you have used the `smtplib` and `email` libraries in the past, you may of found how difficult it is getting the email logic set up properly to handle attachments and different sending scenarios. All customisations are built into Emailee, which takes the guess work out of how to configure.

The following methods are exposed to the user:

#### Emailee.sender(sender: str, replyTo: str = "") -> None

* **sender** - email address of who is sending the email
* **replyTo** (optional) - alternative email address the receiver can reply to

#### Emailee.subject(subject: str = "") -> None

* **subject** (optional) - email subject line, maximum of 255 characters

#### Emailee.msgContent(msgText: str = "", msgHTML: str = "") -> None

* **msgText** (optional) - email message in raw text format
* **msgHTML** (optional) - email message in email HTML format

It's a good idea to use both of these to replicate your email, so users who can't read HTML emails can read your fallback raw text email. Look online for HTML email generators

#### Emailee.sendTo(to: List[str] = [], cc: List[str] = [], bcc: List[str] = [], ignoreErrors: bool = False) -> None

* **to** (optional) - list of to addresses to send emails to
* **cc** (optional) - list of cc addresses to send emails to
* **bcc** (optional) - list of bcc addresses to send emails to
* **ignoreErrors** (optional) - all addresses are validated by regex to be in a valid email format. By default, if an email address is found to be invalid the program will raise an exception. If changed to True, the program will ignore and remove invalid emails and continue with sending the email.

#### Emailee.attachmentFiles(attachmentFiles: List[str] = []) -> None

* **attachmentFiles** (optional) - list of attachments for email, can be listed by relative or full path

#### Emailee.server(smtpServer: str, port: int = 0, SSLTLS: str = "", authUsername: str = "", authPassword: str = "", timeout: int = 30) -> None

* **smtpServer** - server name or IP address of SMTP server
* **port** (optional) - port number for SMTP server, if connection not authenticated and port left blank, port will default to 25. If connection set to SSL or TLS and port left blank, port will default to 465 or 587 respectively.
* **SSLTLS** (optional) - encrypted connection setting, value either "SSL" or "TLS"
* **authUsername** (optional) - username for logging into the SMTP server, if authentication required. If left blank, but authPassword filled in, this will default to the sender email.
* **authPassword** (optional) - password for logging into the SMTP server
* **timeout** (optional) - server connection and email send timeout limit. Extend this value if on a slow network, or sending emails with large attachments

#### Emailee.ready() -> bool

Returns True if the minimal fields to send an email have been set.

#### Email.send(outputFile: str) -> bool

* **outputFile** (optional) - relative or full path location of a file to write report output metadata to, required for async classes when sending mail, file must be empty or not currently exist

Returns True if successful, does not guarantee email was delivered, just sent.

#### Email.__repr__() -> str

Print class function, will print a dictionary of email metadata, useful for testing and debugging.

---

### Data for async classes

Both async classes are very similar in their function and both require emails and server config in the exact same format:

```Python
mail_list = [
    {
        'sender': sender # see Emailee.sender()
        'replyTo': replyTo # see Emailee.sender()
        'subject': email_subject # see Emailee.subject()
        'msgText': raw_text_message # see Emailee.msgContent()
        'msgHTML': HTML_formatted_message # see Emailee.msgContent()
        'to': list_of_to_emails # see Emailee.sendTo()
        'cc': list_of_cc_emails # see Emailee.sendTo()
        'bcc': list_of_bcc_emails # see Emailee.sendTo()
        'ignoreErrors': True # see Emailee.sendTo()
        'attachmentFiles': list_of_attachment_file_paths # see Emailee.attachmentFiles()
    },
    {...},
]
```

```Python
server_dict = {
    'smtpServer': SMTP_server # see Emailee.server()
    'port': port_number # see Emailee.server()
    'SSLTLS': 'TLS' # see Emailee.server()
    'authUsername': authenticated_username # see Emailee.server()
    'authPassword': authenticated_password # see Emailee.server()
}
```

### AsyncThreads class

AsyncThreads uses the threading API to enable asyncronous sending of email. Threading only utilises the same CPU core that the Python program is currently running on, so it can only maximise a single core usage, but the benefits are you can throttle how many threads are allowed to run concurrently and it has a lower overhead compared to `AsyncMP`.

#### AsyncThreads(mailList: List[Dict], serverDict: Dict, outputFile: str, maxThreads: int = 10, waitTime: float = 0)

* **mailList** - see mail_list above for format
* **serverDict** - see server_dict above for format
* **outputFile** - relative or full path location of a file to write report output metadata to, file must be empty or not currently exist. Mandatory for async classes to handle errors.
* **maxThreads** - maximum number of concurrent threads to send emails on
* **waitTime** - wait time in seconds to pause between each thread being spawned, this helps if your SMTP server throttles your connection if you send too many emails in a short time span

### AsyncMP class

AsyncMP uses the multiprocessing API to enable asyncronous sending of email. Multiprocessing utilises all CPU cores of the system the Python program is running on, but does not allow throttling, so you run the risk of maxing out your CPU and RAM, slowing down your system.

#### AsyncMP(mailList: List[Dict], serverDict: Dict, outputFile: str, waitTime: float = 0)

* **mailList** - see mail_list above for format
* **serverDict** - see server_dict above for format
* **outputFile** - relative or full path location of a file to write report output metadata to, file must be empty or not currently exist. Mandatory for async classes to handle errors.
* **waitTime** - wait time in seconds to pause between each process being spawned, this helps if your SMTP server throttles your connection if you send too many emails in a short time span

### Reporting on async output

Upon completion of either async class, you can call the `emailReport()` method to return a metadata list of all emails sent.
This won't work if an exception is thrown during the sending process, hence the **outputFile** requirement.
The output file will be written to after each successful email send so you can analyse against `emailReport()` or see who received emails before a thrown exception or manual cancellation of the sending process.

## Development and Testing

Emailee is built with [**poetry**](https://python-poetry.org/), tested with [**pytest**](https://pytest.org), [**tox**](https://tox.readthedocs.io/en/latest/) and [**coverage**](https://coverage.readthedocs.io/en/coverage-5.5/), type checked with [**mypy**](http://mypy-lang.org/) and formatted with [**black**](https://github.com/psf/black).

### Dev Installation

#### Install with Poetry

Clone the repo, `cd` into it and run `poetry install`

#### Install with pip

Clone the repo, `cd` into it, build and run a new virtual environment, then open the `pyproject.toml` file and `pip install` all packages listed under **[tool.poetry.dev-dependencies]**

### Testing

To successfully run local tests you will need to rename `tests/example.test.env.toml` to `tests/test.env.toml` and modify the config inside the file to contain valid email and SMTP server connection data.

Testing and coverage can then be run with `pytest --cov-report term-missing --cov=emailee tests`

Tox can be used to run pytest against all support Python environments. Open `tox.ini` and check if you have all versions of Python installed listed under *envlist*.

### Pre-commit hooks

Run `pre-commit install` to install the pre-commit hooks in the `.pre-commit-config.yaml` file. Then run `pre-commit run --all-files` to auto-check every file for issues.

Feel free to send me any changes or feedback.

---

Thanks for reading :)

:envelope: :envelope: :envelope: :envelope: :envelope:


CHANGELOG.md
# Changelog

## v1.0.0 (2021-04-24)

Initial stable version release of Emailee

* please try it out, hope it helps!


tests/conftest.py
import os
from pathlib import Path

import pytest
import toml

# generate test environment settings from file if testing locally,
#  otherwise they will be generated from secrets in a github action
if Path("tests/.test.env.toml").is_file():
    try:
        test_config = toml.load("tests/.test.env.toml")
        os.environ["EMAILEE_TEST_SMTP_SERVER"] = test_config["smtp"]["smtp_server"]
        os.environ["EMAILEE_TEST_AUTH_USERNAME"] = test_config["smtp"]["auth_username"]
        os.environ["EMAILEE_TEST_AUTH_PASSWORD"] = test_config["smtp"]["auth_password"]

        os.environ["EMAILEE_TEST_EMAILEE_PATH"] = test_config["variables"][
            "emailee_path"
        ]
        os.environ["EMAILEE_TEST_SENDER"] = test_config["variables"]["sender"]
        os.environ["EMAILEE_TEST_RECEIVER1"] = test_config["variables"]["receiver1"]
        os.environ["EMAILEE_TEST_RECEIVER2"] = test_config["variables"]["receiver2"]
        os.environ["EMAILEE_TEST_RECEIVER3"] = test_config["variables"]["receiver3"]
    except Exception as error:
        raise (error)


# used for numbering the emails in the subject line
def pytest_configure():
    pytest.SENT_EMAIL_NUM = 1
    pytest.SENT_EMAIL_MAX = 36


# delete each async test email output before the next one runs
@pytest.fixture()
def file_setup():
    # Path(os.environ['EMAILEE_TEST_EMAILEE_PATH'] + "/tests/test_output.txt").touch()
    yield
    if Path(
        os.environ["EMAILEE_TEST_EMAILEE_PATH"] + "/tests/test_output.txt"
    ).is_file():
        Path(
            os.environ["EMAILEE_TEST_EMAILEE_PATH"] + "/tests/test_output.txt"
        ).unlink()


# delete all environment variables at end of testing
@pytest.fixture(scope="session", autouse=True)
def teardown():
    yield
    del os.environ["EMAILEE_TEST_SMTP_SERVER"]
    del os.environ["EMAILEE_TEST_AUTH_USERNAME"]
    del os.environ["EMAILEE_TEST_AUTH_PASSWORD"]

    del os.environ["EMAILEE_TEST_EMAILEE_PATH"]
    del os.environ["EMAILEE_TEST_SENDER"]
    del os.environ["EMAILEE_TEST_RECEIVER1"]
    del os.environ["EMAILEE_TEST_RECEIVER2"]
    del os.environ["EMAILEE_TEST_RECEIVER3"]


tests/mail_examples.py
# utilise these test emails or override with your own

textExample: str = """
This is a text email example, \n
You need to add backslash n's to create newlines in a text email.\n\n
Regards\n\n
John Smith
"""

htmlExample: str = """
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-GB">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  <title>Demystifying Email Design</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0"/>

  <style type="text/css">
    a[x-apple-data-detectors] {color: inherit !important;}
  </style>

</head>
<body style="margin: 0; padding: 0;">
  <table role="presentation" border="0" cellpadding="0" cellspacing="0" width="100%">
    <tr>
      <td style="padding: 20px 0 30px 0;">

<table align="center" border="0" cellpadding="0" cellspacing="0" width="600" style="border-collapse: collapse; border: 1px solid #cccccc;">
  <tr>
    <td align="center" bgcolor="#70bbd9" style="padding: 40px 0 30px 0;">
      <img src="https://assets.codepen.io/210284/h1_1.gif" alt="Creating Email Magic." width="300" height="230" style="display: block;" />
    </td>
  </tr>
  <tr>
    <td bgcolor="#ffffff" style="padding: 40px 30px 40px 30px;">
      <table border="0" cellpadding="0" cellspacing="0" width="100%" style="border-collapse: collapse;">
        <tr>
          <td style="color: #153643; font-family: Arial, sans-serif;">
            <h1 style="font-size: 24px; margin: 0;">Lorem ipsum dolor sit amet!</h1>
          </td>
        </tr>
        <tr>
          <td style="color: #153643; font-family: Arial, sans-serif; font-size: 16px; line-height: 24px; padding: 20px 0 30px 0;">
            <p style="margin: 0;">Lorem ipsum dolor sit amet, consectetur adipiscing elit. In tempus adipiscing felis, sit amet blandit ipsum volutpat sed. Morbi porttitor, eget accumsan dictum, nisi libero ultricies ipsum, in posuere mauris neque at erat.</p>
          </td>
        </tr>
        <tr>
          <td>
            <table border="0" cellpadding="0" cellspacing="0" width="100%" style="border-collapse: collapse;">
              <tr>
                <td width="260" valign="top">
                  <table border="0" cellpadding="0" cellspacing="0" width="100%" style="border-collapse: collapse;">
                    <tr>
                      <td>
                        <img src="https://assets.codepen.io/210284/left_1.gif" alt="" width="100%" height="140" style="display: block;" />
                      </td>
                    </tr>
                    <tr>
                      <td style="color: #153643; font-family: Arial, sans-serif; font-size: 16px; line-height: 24px; padding: 25px 0 0 0;">
                        <p style="margin: 0;">Lorem ipsum dolor sit amet, consectetur adipiscing elit. In tempus adipiscing felis, sit amet blandit ipsum volutpat sed. Morbi porttitor, eget accumsan dictum, nisi libero ultricies ipsum, in posuere mauris neque at erat.</p>
                      </td>
                    </tr>
                  </table>
                </td>
                <td style="font-size: 0; line-height: 0;" width="20">&nbsp;</td>
                <td width="260" valign="top">
                  <table border="0" cellpadding="0" cellspacing="0" width="100%" style="border-collapse: collapse;">
                    <tr>
                      <td>
                        <img src="https://assets.codepen.io/210284/right_1.gif" alt="" width="100%" height="140" style="display: block;" />
                      </td>
                    </tr>
                    <tr>
                      <td style="color: #153643; font-family: Arial, sans-serif; font-size: 16px; line-height: 24px; padding: 25px 0 0 0;">
                        <p style="margin: 0;">Lorem ipsum dolor sit amet, consectetur adipiscing elit. In tempus adipiscing felis, sit amet blandit ipsum volutpat sed. Morbi porttitor, eget accumsan dictum, nisi libero ultricies ipsum, in posuere mauris neque at erat.</p>
                      </td>
                    </tr>
                  </table>
                </td>
              </tr>
            </table>
          </td>
        </tr>
      </table>
    </td>
  </tr>
  <tr>
    <td bgcolor="#ee4c50" style="padding: 30px 30px;">
        <table border="0" cellpadding="0" cellspacing="0" width="100%" style="border-collapse: collapse;">
        <tr>
          <td style="color: #ffffff; font-family: Arial, sans-serif; font-size: 14px;">
            <p style="margin: 0;">&reg; Someone, somewhere 2025<br/>
           <a href="#" style="color: #ffffff;">Unsubscribe</a> to this newsletter instantly</p>
          </td>
          <td align="right">
            <table border="0" cellpadding="0" cellspacing="0" style="border-collapse: collapse;">
              <tr>
                <td>
                  <a href="http://www.twitter.com/">
                    <img src="https://assets.codepen.io/210284/tw.gif" alt="Twitter." width="38" height="38" style="display: block;" border="0" />
                  </a>
                </td>
                <td style="font-size: 0; line-height: 0;" width="20">&nbsp;</td>
                <td>
                  <a href="http://www.twitter.com/">
                    <img src="https://assets.codepen.io/210284/fb.gif" alt="Facebook." width="38" height="38" style="display: block;" border="0" />
                  </a>
                </td>
              </tr>
            </table>
          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>

      </td>
    </tr>
  </table>
</body>
</html>
"""


emailee/emailee.py
import enum
import mimetypes
import re
import smtplib
from email.mime.application import MIMEApplication
from email.mime.audio import MIMEAudio
from email.mime.image import MIMEImage
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.utils import formatdate
from pathlib import Path
from typing import BinaryIO, List, TextIO, Union


def _helperOutputFileCheck(filename: str) -> bool:
    try:
        if Path(filename).is_file():
            try:
                fileSize: int = Path(filename).stat().st_size
            except Exception as error:
                raise error
            if fileSize != 0:
                raise OSError(
                    "outputFile is not empty, please empty or delete it and try again"
                )
        else:
            try:
                Path(filename).touch()
            except Exception as error:
                raise PermissionError(f"No access to create the new file - {error}")
    except Exception as error:
        raise PermissionError(
            f"No access to file location to check if file exists - {error}"
        )
    return True


class _EncType(enum.Enum):
    """
    Enumerations of TLS and SSL, for encryption logic in Emailee
    """

    SSL = 1
    TLS = 2


class Emailee:
    """
    A robust SMTP email solution for Python.
    Zero external dependencies (unless testing).
    Allows sending via SSL or TLS, attachments, reply-to and to/cc/bcc.
    Emails are built with the below structure -

    multipart/mixed
        ├─── multipart/related
        │    └── multipart/alternative
        │        ├─── text/plain
        │        └─── text/html
        ├─── application/text - embedded attachment
        ├─── application/image - embedded attachment
        ├─── application/audio - embedded attachment
        └─── application/pdf etc. - other application attachment

    See https://github.com/mattwalshdev/emailee for additional documentation

    Methods
    -------
        sender(sender: str, replyTo: str) - email (required), email
        subject(subject: str) - text
        msgContent(msgText: str, msgHTML: str) - raw text, html text
        sendTo(to: List[str], cc: List[str], bcc: List[str], ignoreErrors: bool) -
            List of emails in each, one required. Set ignoreErrors True if you
            want it to still send emails if there is 1+ invalid emails listed
        attachmentFiles(attachmentFiles: List[str]) - List of full and/or
            relative paths to files
        server(smtpServer: str, port: int, SSLTLS: str, authUsername: str,
            authPassword: str, timeout: int) - servername (required),
            port no, encryption type, credentials,
            timeout for server connection timeout in seconds, default 30 seconds,
            increase if on slow connection
        ready() - will return True if all required fields are entered and valid,
            does not guarantee emails actually exist
        send(outputFile: str) - attempt to send the email, optional addition of an
            outputFile to save metadata to. OutputFile is mandatory when sending emails in either async mode

    Example
    -------
    import emailee

    my_email = emailee.Emailee()
    my_email.sender("john.smith@fakeemail.com")
    my_email.subject("Test email")
    my_email.msgContent("raw text, poorly formatted")
    my_email.sendTo(["jill.smith@fakeemail.com"])
    my_email.server("smtp.fakeemail.com")

    if my_email.ready():
        my_email.send()
    """

    _EMAIL_REGEX = re.compile(r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)")

    def __init__(self) -> None:
        self._sender: str = ""
        self._replyTo: str = ""
        self._subject: str = ""
        self._msgText: str = ""
        self._msgHTML: str = ""
        self._to: List[str] = []
        self._cc: List[str] = []
        self._bcc: List[str] = []
        self._attachmentFiles: List[str] = []
        self._smtpServer: str = ""
        self._port: int = 0
        self._SSLTLS: Union[_EncType, None] = None
        self._authUsername: str = ""
        self._authPassword: str = ""
        self._timeout: int = 30
        self._outputFile: str = ""
        self._outputFileReady: bool = False

    def __repr__(self) -> str:
        outputDict = {
            "sender": self._sender,
            "replyTo": self._replyTo,
            "subject": self._subject,
            "msgText": self._msgText,
            "msgHTML": self._msgHTML,
            "to": self._to,
            "cc": self._cc,
            "bcc": self._bcc,
            "attachmentFiles": self._attachmentFiles,
            "smtpServer": self._smtpServer,
            "port": self._port,
            "SSLTLS": self._SSLTLS,
        }
        return str(outputDict)

    def __str__(self) -> str:
        outputDict = {
            "sender": self._sender,
            "replyTo": self._replyTo,
            "subject": self._subject,
            "to": self._to,
            "cc": self._cc,
            "bcc": self._bcc,
        }
        return str(outputDict)

    def _addressCheck(
        self, emailList: List[str], type: str, ignoreErrors: bool
    ) -> List[str]:
        self._tempEmailList = []
        for email in emailList:
            if not isinstance(email, str):
                raise TypeError(f"{type} email not a string - {email}")

            if not Emailee._EMAIL_REGEX.fullmatch(email):
                if not ignoreErrors:
                    raise ValueError(f"Invalid {type} email - {email}")
                else:
                    continue
            if ignoreErrors:
                self._tempEmailList.append(email)
        if ignoreErrors:
            return self._tempEmailList
        return emailList

    def sender(self, sender: str, replyTo: str = "") -> None:
        if not isinstance(sender, str):
            raise TypeError("sender not a string")

        self._sender = self._addressCheck([sender], "sender", ignoreErrors=False)[0]

        if replyTo:
            if not isinstance(replyTo, str):
                raise TypeError("replyTo not a string")

            self._replyTo = self._addressCheck(
                [replyTo], "replyTo", ignoreErrors=False
            )[0]

    def subject(self, subject: str) -> None:
        if not isinstance(subject, str):
            raise TypeError("subject not a string")

        if len(subject) > 255:
            raise ValueError("subject string too long")
        self._subject = subject

    def msgContent(self, msgText: str = "", msgHTML: str = "") -> None:
        if not isinstance(msgText, str):
            raise TypeError("msgText not a string")

        if not isinstance(msgHTML, str):
            raise TypeError("msgHTML not a string")

        self._msgText = msgText
        self._msgHTML = msgHTML

    def sendTo(
        self,
        to: List[str] = [],
        cc: List[str] = [],
        bcc: List[str] = [],
        ignoreErrors: bool = False,
    ) -> None:
        if not to and not cc and not bcc:
            raise ValueError("No email addresses to send to are defined")

        if (
            not isinstance(to, list)
            or not isinstance(cc, list)
            or not isinstance(bcc, list)
        ):
            raise TypeError("sendTo items are not in list format")

        if not isinstance(ignoreErrors, bool):
            raise TypeError("sendTo ignoreErrors is not a bool")

        self._to = self._addressCheck(to, "to", ignoreErrors)
        self._cc = self._addressCheck(cc, "cc", ignoreErrors)
        self._bcc = self._addressCheck(bcc, "bcc", ignoreErrors)

        if not self._to and not self._cc and not self._bcc:
            raise ValueError("No email addresses to send to are properly defined")

    def attachmentFiles(self, attachmentFiles: List[str] = []) -> None:
        if not isinstance(attachmentFiles, list):
            raise TypeError("attachmentFiles not in list format")

        for file in attachmentFiles:
            if not isinstance(file, str):
                raise TypeError(f"file name listed not in string format - {file}")

            if not Path(file).is_file():
                raise ValueError(
                    f"Attachment listed is not a file or not accessible - {file}"
                )

        self._attachmentFiles = attachmentFiles

    def server(
        self,
        smtpServer: str,
        *,
        port: int = 0,
        SSLTLS: str = "",
        authUsername: str = "",
        authPassword: str = "",
        timeout: int = 30,
    ) -> None:
        if not isinstance(smtpServer, str):
            raise TypeError("smtpServer name not in string format")

        if not isinstance(port, int):
            raise TypeError("smtpServer port not in integer format")

        if not isinstance(SSLTLS, str):
            raise TypeError("smtpServer SSLTLS option not in string format")

        if not isinstance(authUsername, str):
            raise TypeError("smtpServer auth username not in string format")

        if not isinstance(authPassword, str):
            raise TypeError("smtpServer auth password not in string format")

        if not isinstance(timeout, int):
            raise TypeError("timeout server option not in integer format")

        if port < 0 or port > 65535:
            raise ValueError("Port is an invalid number")

        if SSLTLS not in _EncType.__members__ and SSLTLS != "":
            raise ValueError("SSLTLS option is invalid")

        if timeout <= 0:
            raise ValueError("Server timeout period is an invalid number")

        self._smtpServer = smtpServer

        if port > 0:
            self._port = port
        elif SSLTLS == "TLS":
            self._port = 587
        elif SSLTLS == "SSL":
            self._port = 465
        else:
            self._port = 25

        if SSLTLS:
            self._SSLTLS = _EncType[f"{SSLTLS}"]
        elif port == 587:
            self._SSLTLS = _EncType.TLS
        elif port == 465:
            self._SSLTLS = _EncType.SSL

        if authUsername and authPassword:
            self._authUsername = authUsername
            self._authPassword = authPassword
        elif authPassword:
            self._authUsername = self._sender
            self._authPassword = authPassword

        self._timeout = timeout

    def ready(self) -> bool:
        if self._authPassword and not self._authUsername and self._sender:
            # this may happen if server methods are invoked before
            #  the sender method AND the person has not added an
            #  authUsername BUT has added an authPassword
            self._authUsername = self._sender

        if (
            self._sender
            and (self._to or self._cc or self._bcc)
            and self._smtpServer
            and self._port
        ):
            # minimum requirements in place to send an email
            return True
        else:
            return False

    def _generate(self) -> str:
        message = MIMEMultipart("mixed")

        message["to"] = ", ".join(self._to) if self._to else ""

        message["cc"] = ", ".join(self._cc) if self._cc else ""

        message["from"] = self._sender
        message["subject"] = self._subject
        message["date"] = formatdate(localtime=True)

        if self._replyTo:
            message.add_header("reply-to", self._replyTo)

        messageRelated = MIMEMultipart("related")
        messageAlternative = MIMEMultipart("alternative")

        if self._msgText:
            messageAlternative.attach(MIMEText(self._msgText, "plain"))
        if self._msgHTML:
            messageAlternative.attach(MIMEText(self._msgHTML, "html"))

        messageRelated.attach(messageAlternative)

        message.attach(messageRelated)

        for attFile in self._attachmentFiles or []:
            contentType, encoding = mimetypes.guess_type(attFile)

            if contentType is None or encoding is not None:
                contentType = "application/octet-stream"
            mainType, subType = contentType.split("/", 1)

            attachment: Union[MIMEText, MIMEImage, MIMEAudio, MIMEApplication]
            fp: Union[TextIO, BinaryIO]

            if mainType == "text":
                with open(attFile, "r") as fp:
                    attachment = MIMEText(fp.read(), _subtype=subType)
            else:
                with open(attFile, "rb") as fp:
                    if mainType == "image":
                        attachment = MIMEImage(fp.read(), _subtype=subType)
                    elif mainType == "audio":
                        attachment = MIMEAudio(fp.read(), _subtype=subType)
                    else:
                        attachment = MIMEApplication(fp.read(), _subtype=subType)

            filename = Path(attFile).stem
            attachment.add_header(
                "Content-Disposition", "attachment; filename=" + filename
            )
            message.attach(attachment)

        return message.as_string()

    def send(self, outputFile: str = "") -> bool:
        if self._authPassword and not self._authUsername and self._sender:
            # this may happen if server methods are invoked before
            #  the sender method AND the person has not added an
            #  authUsername BUT has added an authPassword
            self._authUsername = self._sender

        if not isinstance(outputFile, str):
            raise TypeError("outputFile path not in string format")

        # run checks against the output file, unless it's already been run
        #  during this run, this is for async batch jobs of emails
        #  where send is called repeatedly
        if outputFile:
            if not self._outputFileReady:
                self._outputFileReady = _helperOutputFileCheck(outputFile)
            self._outputFile = outputFile

        try:
            generatedEmail: str = self._generate()
        except Exception as error:
            raise (error)

        try:
            smtp: object

            if self._SSLTLS == _EncType.SSL:
                try:
                    smtp = smtplib.SMTP_SSL(
                        self._smtpServer, self._port, timeout=self._timeout
                    )
                except Exception as error:
                    raise ValueError(error)
            else:
                try:
                    smtp = smtplib.SMTP(
                        self._smtpServer, self._port, timeout=self._timeout
                    )
                except Exception as error:
                    raise ValueError(error)

            smtp.ehlo()

            if self._SSLTLS == _EncType.TLS:
                smtp.starttls()
                smtp.ehlo()
        except Exception as error:
            raise error
        if self._authUsername and self._authPassword:
            try:
                smtp.login(self._authUsername, self._authPassword)
            except Exception as error:
                raise ValueError(error)

        try:
            smtp.sendmail(self._sender, self._to + self._cc + self._bcc, generatedEmail)
        except Exception as error:
            raise ValueError(error)

        try:
            smtp.close()
            if self._outputFile:
                with open(self._outputFile, "a") as file:
                    file.write(self.__str__() + "\n")
            return True
        except Exception as error:
            raise (error)
